src/contactsmodel.cpp

```diff
@@ -0,0 +1,187 @@
+        this->canUseDeviceContacts = true;
+    } else {
+        LOG("Error opening device's contacts database :(");
+        this->canUseDeviceContacts = false;
+    }
+}
+
+int ContactsModel::rowCount(const QModelIndex &) const
+{
+    return this->filter.isEmpty() ? this->contacts.size() : this->filteredContacts.size();
+}
+
+QVariant ContactsModel::data(const QModelIndex &index, int role) const
+{
+    if (index.isValid() && role == Qt::DisplayRole) {
+        return this->filter.isEmpty() ? QVariant(contacts.value(index.row())) : QVariant(filteredContacts.value(index.row())) ;
+    }
+    return QVariant();
+}
+
+void ContactsModel::handleUsersReceived(const QString &extra, const QVariantList &userIds, int totalUsers)
+{
+    if (extra == "contactsRequested") {
+        LOG("Received contacts list..." << totalUsers);
+        this->contactIds.clear();
+        QListIterator<QVariant> userIdIterator(userIds);
+        while (userIdIterator.hasNext()) {
+            QString nextUserId = userIdIterator.next().toString();
+            if (!this->tdLibWrapper->hasUserInformation(nextUserId)) {
+                this->tdLibWrapper->getUserFullInfo(nextUserId);
+            }
+            this->contactIds.append(nextUserId);
+        }
+    }
+}
+
+static bool compareUsers(const QVariant &user1, const QVariant &user2)
+{
+    const QVariantMap userMap1 = user1.toMap();
+    const QVariantMap userMap2 = user2.toMap();
+
+    const QString lastName1 = userMap1.value(LAST_NAME).toString();
+    const QString lastName2 = userMap2.value(LAST_NAME).toString();
+    if (!lastName1.isEmpty()) {
+        if (lastName1 < lastName2) {
+            return true;
+        } else if (lastName1 > lastName2) {
+            return false;
+        }
+    }
+
+    const QString firstName1 = userMap1.value(FIRST_NAME).toString();
+    const QString firstName2 = userMap2.value(FIRST_NAME).toString();
+    if (firstName1 < firstName2) {
+        return true;
+    } else if (firstName1 > firstName2) {
+        return false;
+    }
+    const QString userName1 = userMap1.value(USERNAME).toString();
+    const QString userName2 = userMap2.value(USERNAME).toString();
+    if (userName1 < userName2) {
+        return true;
+    } else if (userName1 > userName2) {
+        return false;
+    }
+    return userMap1.value(ID).toLongLong() < userMap2.value(ID).toLongLong();
+}
+
+void ContactsModel::hydrateContacts()
+{
+    LOG("Hydrating contacts...");
+    this->contacts.clear();
+    QListIterator<QString> userIdIterator(contactIds);
+    while (userIdIterator.hasNext()) {
+        QString nextUserId = userIdIterator.next();
+        LOG("Hydrating contact:" << nextUserId);
+        this->contacts.append(this->tdLibWrapper->getUserInformation(nextUserId));
+    }
+    LOG("Hydrated contacts:" << this->contacts.size());
+    std::sort(this->contacts.begin(), this->contacts.end(), compareUsers);
+}
+
+void ContactsModel::applyFilter(const QString &filter)
```

Comment:
Consider using `QSortFilterProxyModel` for filtering and sorting. AFAICT it does a pretty good job in avoiding unnecessary updates of the view (let alone resetting the model on every change).

Reply:
Initially I wanted to implement it with `QSortFilterProxyModel`, but no matter what I did (applied a filter or not), I always ended up with an empty list. Maybe I'm simply too dumb to understand or once again used it in the wrong way and got stuck because of that. Anyway, I gave up and implemented it that way...

Reply:
Trust me, it does work :) If matches only one role though - if you want it to match several roles, then you would need to define additional "combined" role (not displayed in the UI) which would concatenate all roles involved in matching into one (separated with `'\n'` or something). But that obviously contradicts the "one role fits all" and "why would anyone need more than one column in a database" approach.

Reply:
I went for a ride with the bike and for whatever reason it worked straight afterwards - though I'm close to 100% sure that I didn't do anything different than when I tried it last a few days ago. Anyway, we're using the `QSortFilterProxyModel` and I removed the duplicate filtered list...